File: app/src/main/java/hibernate/v2/testyourandroid/ui/fragment/MonitorCpuFragment.java
package hibernate.v2.testyourandroid.ui.fragment;

import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.jjoe64.graphview.GraphView;
import com.jjoe64.graphview.GraphView.GraphViewData;
import com.jjoe64.graphview.GraphViewSeries;
import com.jjoe64.graphview.GraphViewSeries.GraphViewSeriesStyle;
import com.jjoe64.graphview.GraphViewStyle;
import com.jjoe64.graphview.LineGraphView;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileFilter;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import hibernate.v2.testyourandroid.R;

public class MonitorCpuFragment extends BaseFragment {

	private final int CPU_MIN = 0;
	private final int CPU_MAX = 1;
	private final int CPU_CURRENT = 2;

	private GraphViewSeries series;
	private double lastXValue = 0;

	private Handler mHandler = new Handler();

	private Runnable timer = new Runnable() {
		@Override
		public void run() {
			lastXValue += 1;
			speedText.setText(getCPU(CPU_CURRENT) + " MHz");
			series.appendData(new GraphViewData(lastXValue, getCPU(CPU_CURRENT)),
					true, 100);
			mHandler.postDelayed(timer, 1000);
		}
	};

	@BindView(R.id.graph2)
	LinearLayout layout;
	@BindView(R.id.coreText)
	TextView coreText;
	@BindView(R.id.minText)
	TextView minText;
	@BindView(R.id.maxText)
	TextView maxText;
	@BindView(R.id.speedText)
	TextView speedText;

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
	                         Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_monitor_cpu, container, false);
		ButterKnife.bind(this, view);
		return view;
	}

	@Override
	public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		speedText.setText(getCPU(CPU_CURRENT) + " MHz");
		coreText.setText(String.valueOf(getNumCores()));
		minText.setText(getCPU(CPU_MIN) + " MHz");
		maxText.setText(getCPU(CPU_MAX) + " MHz");
		GraphView graphView = new LineGraphView(mContext, "");
		graphView.setShowHorizontalLabels(false);
		GraphViewData[] data = new GraphViewData[]{new GraphViewData(0, 0)};
		series = new GraphViewSeries("", new GraphViewSeriesStyle(
				Color.parseColor("#33B5E5"), 3), data);
		graphView.addSeries(series);
		graphView.setGraphViewStyle(new GraphViewStyle(Color.GRAY, Color.GRAY,
				Color.LTGRAY));
		graphView.setManualYAxisBounds(getCPU(CPU_MAX), 0);
		graphView.setViewPort(0, 36);
		graphView.setScalable(true);
		graphView.setScrollable(true);
		graphView.setDisableTouch(true);
		layout.addView(graphView);
	}

	@Override
	public void onResume() {
		super.onResume();
		mHandler.postDelayed(timer, 1000);
	}

	@Override
	public void onPause() {
		super.onPause();
		mHandler.removeCallbacks(timer);
	}

	private int getCPU(int type) {
		String filename = "";
		if (type == CPU_MIN) {
			filename = "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_min_freq";
		} else if (type == CPU_MAX) {
			filename = "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq";
		} else if (type == CPU_CURRENT) {
			filename = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq";
		}
		ArrayList<String> list = new ArrayList<>();
		if (new File(filename).exists()) {
			try {
				BufferedReader br = new BufferedReader(new FileReader(new File(
						filename)));
				String aLine;
				while ((aLine = br.readLine()) != null) {
					list.add(aLine);
				}
				br.close();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
		StringBuilder sd = new StringBuilder();
		for (String e : list) {
			sd.append(e);
		}
		try {
			return Integer.valueOf(sd.toString()) / 1000;
		} catch (NumberFormatException e) {
			return 0;
		}
	}

	private int getNumCores() {
		// Private Class to display only CPU devices in the directory listing
		class CpuFilter implements FileFilter {
			@Override
			public boolean accept(File pathname) {
				// Check if filename is "cpu", followed by a single digit number
				return Pattern.matches("cpu[0-9]", pathname.getName());
			}
		}
		try {
			// Get directory containing CPU info
			File dir = new File("/sys/devices/system/cpu/");
			// Filter to only list the devices we care about
			File[] files = dir.listFiles(new CpuFilter());
			// Return the number of cores (virtual CPU devices)
			return files.length;
		} catch (Exception e) {
			// Default to return 1 core
			return 1;
		}
	}
}

File: app/src/main/java/hibernate/v2/testyourandroid/ui/info/monitor/MonitorCPUFragment.kt
package hibernate.v2.testyourandroid.ui.info.monitor

import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import com.blankj.utilcode.util.ConvertUtils
import com.jjoe64.graphview.GridLabelRenderer
import com.jjoe64.graphview.series.DataPoint
import com.jjoe64.graphview.series.LineGraphSeries
import hibernate.v2.testyourandroid.R
import hibernate.v2.testyourandroid.databinding.FragmentMonitorCpuBinding
import hibernate.v2.testyourandroid.ui.base.BaseFragment
import java.io.BufferedReader
import java.io.File
import java.io.FileFilter
import java.io.FileReader
import java.io.IOException
import java.util.ArrayList
import java.util.regex.Pattern

class MonitorCPUFragment : BaseFragment<FragmentMonitorCpuBinding>() {

    override fun getViewBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): FragmentMonitorCpuBinding =
        FragmentMonitorCpuBinding.inflate(inflater, container, false)

    private var series = LineGraphSeries(arrayOf<DataPoint>())
    private var lastXValue = 0.0
    private val mHandler = Handler(Looper.getMainLooper())
    private val timer: Runnable = object : Runnable {
        override fun run() {
            viewBinding?.speedText?.text = "${getCPU(CPU_CURRENT)} MHz"
            series.appendData(DataPoint(lastXValue, getCPU(CPU_CURRENT).toDouble()), true, 36)
            viewBinding?.graphView?.viewport?.scrollToEnd()
            lastXValue += 1.0
            mHandler.postDelayed(this, UPDATE_CHART_INTERVAL)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewBinding?.let { viewBinding ->
            context?.let { context ->
                viewBinding.speedText.text = "${getCPU(CPU_CURRENT)} MHz"
                viewBinding.coreText.text = numCores.toString()
                viewBinding.minText.text = "${getCPU(CPU_MIN)} MHz"
                viewBinding.maxText.text = "${getCPU(CPU_MAX)} MHz"
                series.thickness = ConvertUtils.dp2px(4f)
                series.color = ContextCompat.getColor(context, R.color.lineColor1)
                series.isDrawBackground = true
                series.backgroundColor = ContextCompat.getColor(context, R.color.lineColor1A)
                viewBinding.graphView.addSeries(series)
                viewBinding.graphView.gridLabelRenderer.gridColor = Color.GRAY
                viewBinding.graphView.gridLabelRenderer.isHighlightZeroLines = false
                viewBinding.graphView.gridLabelRenderer.isHorizontalLabelsVisible = false
                viewBinding.graphView.gridLabelRenderer.padding = ConvertUtils.dp2px(10f)
                viewBinding.graphView.gridLabelRenderer.gridStyle =
                    GridLabelRenderer.GridStyle.HORIZONTAL
                viewBinding.graphView.viewport.isYAxisBoundsManual = true
                viewBinding.graphView.viewport.setMinY(0.0)
                viewBinding.graphView.viewport.setMaxY(getCPU(CPU_MAX).toDouble())
                viewBinding.graphView.viewport.isXAxisBoundsManual = true
                viewBinding.graphView.viewport.setMinX(0.0)
                viewBinding.graphView.viewport.setMaxX(36.0)
                viewBinding.graphView.viewport.isScrollable = false
                viewBinding.graphView.viewport.isScalable = false
            }
        }
    }

    override fun onResume() {
        super.onResume()
        mHandler.postDelayed(timer, UPDATE_CHART_INTERVAL)
    }

    override fun onPause() {
        super.onPause()
        mHandler.removeCallbacks(timer)
    }

    private fun getCPU(type: Int): Int {
        var filename = ""
        when (type) {
            CPU_MIN -> filename = "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_min_freq"
            CPU_MAX -> filename = "/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq"
            CPU_CURRENT -> filename = "/sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq"
        }
        val list = ArrayList<String>()
        val file = File(filename)
        if (file.exists()) {
            try {
                val bufferedReader = BufferedReader(FileReader(file))
                for (line in bufferedReader.readLine()) {
                    list.add(line.toString())
                }
                bufferedReader.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        val sd = StringBuilder()
        for (e in list) {
            sd.append(e)
        }
        return try {
            Integer.valueOf(sd.toString()) / 1000
        } catch (e: NumberFormatException) {
            0
        }
    }

    // Default to return 1 core
    // Get directory containing CPU info
    // Filter to only list the devices we care about
    // Return the number of cores (virtual CPU devices)
    // Check if filename is "cpu", followed by a single digit number
    // Private Class to display only CPU devices in the directory listing
    private val numCores: Int
        get() { // Private Class to display only CPU devices in the directory listing
            class CpuFilter : FileFilter {
                override fun accept(pathname: File): Boolean { // Check if filename is "cpu", followed by a single digit number
                    return Pattern.matches("cpu[0-9]", pathname.name)
                }
            }
            return try { // Get directory containing CPU info
                val dir = File("/sys/devices/system/cpu/")
                // Filter to only list the devices we care about
                val files = dir.listFiles(CpuFilter())
                // Return the number of cores (virtual CPU devices)
                files?.size ?: 1
            } catch (e: Exception) { // Default to return 1 core
                1
            }
        }

    companion object {
        private const val CPU_MIN = 0
        private const val CPU_MAX = 1
        private const val CPU_CURRENT = 2
        const val UPDATE_CHART_INTERVAL = 1000L
    }
}